src/analysis/normed/group/seminorm.lean

```diff
@@ -51,44 +51,88 @@ structure group_seminorm (G : Type*) [group G] :=
 
 attribute [nolint doc_blame] add_group_seminorm.to_zero_hom
 
+/-- `add_group_seminorm_class F α` states that `F` is a type of seminorms on the additive group `α`.
+
+You should extend this class when you extend `add_group_seminorm`. -/
+class add_group_seminorm_class (F : Type*) (α : out_param $ Type*) [add_group α]
+  extends fun_like F α (λ _, ℝ) :=
+(map_zero (f : F) : f 0 = 0)
+(map_add_le_add (f : F) (a b : α) : f (a + b) ≤ f a + f b)
+(map_neg_eq_map (f : F) (a : α) : f (-a) = f a)
+
+/-- `group_seminorm_class F α` states that `F` is a type of seminorms on the group `α`.
+
+You should extend this class when you extend `group_seminorm`. -/
+@[to_additive]
+class group_seminorm_class (F : Type*) (α : out_param $ Type*) [group α]
+  extends fun_like F α (λ _,  ℝ) :=
+(map_one_eq_zero (f : F) : f 1 = 0)
+(map_mul_le_add (f : F) (a b : α) : f (a * b) ≤ f a + f b)
+(map_inv_eq_map (f : F) (a : α) : f a⁻¹ = f a)
+
+export add_group_seminorm_class (map_add_le_add map_neg_eq_map)
+export group_seminorm_class (map_one_eq_zero map_mul_le_add map_inv_eq_map)
+
+attribute [simp, to_additive map_zero] map_one_eq_zero
+attribute [simp] map_neg_eq_map
+attribute [simp, to_additive] map_inv_eq_map
+
+@[priority 100] -- See note [lower instance priority]
+instance add_group_seminorm_class.to_zero_hom_class [add_group E] [add_group_seminorm_class F E] :
+  zero_hom_class F E ℝ :=
+{ ..‹add_group_seminorm_class F E› }
```

Comment:
It's probably a good idea to copy over the data-carrying fields, for better unfolding.
```suggestion
{ coe := coe_fn,
  ..‹add_group_seminorm_class F E› }
```